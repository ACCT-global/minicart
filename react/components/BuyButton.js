import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { graphql, compose } from 'react-apollo'
import { FormattedMessage } from 'react-intl'
import Button from '@vtex/styleguide/lib/Button'

import orderFormQuery from '../queries/orderFormQuery.gql'
import addToCartMutation from '../mutations/addToCartMutation.gql'

export class BuyButton extends Component {
  static propTypes = {
    id: PropTypes.string,
    data: PropTypes.object,
    mutate: PropTypes.func,
    onClick: PropTypes.func,
  }

  handleClick = () => {
    const {
      data: { orderForm: { orderFormId } },
      mutate,
      id,
      onClick,
    } = this.props
    mutate({
      variables: {
        orderFormId: orderFormId,
        items: [
          {
            id: parseInt(id),
            index: 1,
            quantity: 1,
            seller: 1,
          },
        ],
      },
      refetchQueries: [{ query: orderFormQuery }],
    })
    onClick()
  }

  render() {
    return (
      <Button primary block large onClick={this.handleClick}>
        <FormattedMessage id="dreamstore.buy" />
      </Button>
    )
  }
}

export default compose(
  graphql(orderFormQuery, {
    options: { ssr: false },
  }),
  graphql(addToCartMutation)
)(BuyButton)
