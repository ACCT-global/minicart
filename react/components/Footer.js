import React, { Component } from 'react'
import { injectIntl, intlShape } from 'react-intl'

class Footer extends Component {
  static propTypes = {
    intl: intlShape.isRequired,
  }

  translate = id => this.props.intl.formatMessage({ id: `dreamstore.${id}` })

  render() {
    const year = new Date().getFullYear()
    return (
      <footer className="w-100 bottom-0 pa7 pa9-l bg-near-white flex items-center">
        <div className="mw8 center">
          <div className="cf tc fw7 flex items-center justify-center">
            {this.translate('poweredBy')}
          </div>
          <div className="tc">{year}</div>
        </div>
      </footer>
    )
  }
}

export default injectIntl(Footer)
