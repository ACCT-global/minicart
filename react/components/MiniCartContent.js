import React, { Component, Fragment } from 'react'
import PropTypes from 'prop-types'
import { injectIntl, intlShape } from 'react-intl'
import { reduceBy, values } from 'ramda'
import classNames from 'classnames'
import { ExtensionPoint } from 'render'
import { Button, Spinner, IconDelete } from 'vtex.styleguide'
import ProductPrice from 'vtex.store-components/ProductPrice'
import { MiniCartPropTypes } from '../propTypes'

/**
 * Minicart content component
 */
class MiniCartContent extends Component {
  static propTypes = {
    /* Set the mini cart content size */
    large: PropTypes.bool,
    /* Internationalization */
    intl: intlShape.isRequired,
    /** Define a function that is executed when the item is clicked */
    actionOnClick: PropTypes.func,
    /* Reused props */
    data: MiniCartPropTypes.orderFormContext,
    labelMiniCartEmpty: MiniCartPropTypes.labelMiniCartEmpty,
    labelButton: MiniCartPropTypes.labelButtonFinishShopping,
    showRemoveButton: MiniCartPropTypes.showRemoveButton,
    showSku: MiniCartPropTypes.showSku,
    enableQuantitySelector: MiniCartPropTypes.enableQuantitySelector,
    maxQuantity: MiniCartPropTypes.maxQuantity,
    showDiscount: MiniCartPropTypes.showDiscount,
  }

  state = { isUpdating: false }

  sumItemsPrice = items => {
    let sum = 0
    items.forEach(item => {
      sum += item.listPrice * item.quantity
    })
    return sum
  }

  getGroupedItems = () =>
    values(
      reduceBy(
        (acc, item) =>
          acc ? { ...acc, quantity: acc.quantity + item.quantity } : item,
        undefined,
        item => item.id,
        this.props.data.orderForm.items
      )
    )

  calculateDiscount = (items, totalPrice) =>
    this.sumItemsPrice(items) - totalPrice

  handleClickButton = () => location.assign('/checkout/#/cart')

  onRemoveItem = id => {
    this.setState({ isUpdating: true })

    const {
      data: { orderForm, updateAndRefetchOrderForm },
    } = this.props
    const itemPayload = orderForm.items.find(item => item.id === id)
    const index = orderForm.items.indexOf(itemPayload)
    const updatedItem = [itemPayload].map(item => {
      return {
        id: parseInt(item.id),
        index: index,
        quantity: 0,
        seller: 1,
      }
    })

    updateAndRefetchOrderForm({
      variables: {
        orderFormId: orderForm.orderFormId,
        items: updatedItem,
      },
    }).then(() => {
      this.setState({ isUpdating: false })
    })
  }

  onUpdateItems = (id, quantity) => {
    this.setState({ isUpdating: true })
    const {
      data: {
        orderForm,
        updateAndRefetchOrderForm,
      },
    } = this.props
    const items = this.getGroupedItems()
    const itemPayloadGrouped = items.find(item => item.id === id)
    const itemsPayload = orderForm.items.filter(item => item.id === id)
    let itemPayload = itemsPayload[0]
    const index = orderForm.items.indexOf(itemsPayload[0])
    const newQuantity = quantity - (itemPayloadGrouped.quantity - itemPayload.quantity)
    const updatedItems = [
      {
        id: itemPayload.id,
        index,
        quantity: newQuantity,
      },
    ]

    if (newQuantity <= 0) {
      updatedItems[0].quantity = 0
      itemPayload = itemsPayload[1]
      updatedItems.push(
        {
          id: itemPayload.id,
          index: orderForm.items.indexOf(itemPayload),
          quantity: itemPayload.quantity + newQuantity,
        }
      )
    }

    updateAndRefetchOrderForm({
      variables: {
        orderFormId: orderForm.orderFormId,
        items: updatedItems,
      },
    }).then(() => {
      this.setState({ isUpdating: false })
    })
  }

  renderWithoutItems = label => (
    <div className="vtex-minicart__item pa9 flex items-center justify-center relative bg-base">
      <span className="t-body">{label}</span>
    </div>
  )

  createProductShapeFromItem = item => ({
    productName: item.name,
    linkText: item.detailUrl.replace(/^\//, '').replace(/\/p$/, ''),
    sku: {
      seller: {
        commertialOffer: {
          Price: item.sellingPrice,
          ListPrice: item.ListPrice
        }
      },
      name: item.skuName,
      itemId: item.id,
      image: {
        imageUrl: item.imageUrl
      },
    },
  })

  renderMiniCartWithItems = (
    orderForm,
    label,
    labelDiscount,
    showRemoveButton,
    showDiscount,
    showSku,
    enableQuantitySelector,
    maxQuantity,
    actionOnClick,
    isUpdating,
    large
  ) => {
    const items = this.getGroupedItems()

    const classes = classNames(
      'vtex-minicart__content overflow-x-hidden pa1',
      {
        'vtex-minicart__content--small bg-base': !large,
        'overflow-y-auto': large,
        'overflow-y-scroll': items.length > 3 && !large,
        'overflow-y-hidden': items.length <= 3 && !large,
      }
    )

    const discount = this.calculateDiscount(items, orderForm.value)

    return (
      <Fragment>
        <div className={classes}>
          {items.map(item => (
            <Fragment key={item.id}>
              <div className="relative flex">
                <div className="fr absolute bottom-0 right-0">
                  <Button icon variation="tertiary" disabled={isUpdating} onClick={e => this.onRemoveItem(item.id)}>
                    <IconDelete size={15} color="silver" />
                  </Button>
                </div>
                <ExtensionPoint id="product-summary"
                  showBorders
                  product={this.createProductShapeFromItem(item)}
                  name={item.name}
                  displayMode="inline"
                  showListPrice={false}
                  showBadge={false}
                  showInstallments={false}
                  showLabels={false}
                  actionOnClick={actionOnClick}
                />
              </div>
            </Fragment>
          ))}
        </div>
        <div className="vtex-minicart-content__footer w-100 bg-base pa4 bt b--muted-3 pt4 flex flex-column items-end">
          {showDiscount && (
            <div className="vtex-minicart__content-discount blue w-100 flex justify-end items-center">
              <span className="ttl c-action-primary">{labelDiscount}</span>
              <ProductPrice
                sellingPrice={discount}
                listPrice={discount}
                showLabels={false}
                showListPrice={false}
              />
            </div>
          )}
          <div className="vtex-minicart__content-price mb3">
            {isUpdating && <Spinner size={18} />}
            <ProductPrice
              sellingPrice={orderForm.value}
              listPrice={orderForm.value}
              showLabels={false}
              showListPrice={false}
            />
          </div>
          <Button
            variation="primary"
            size="small"
            onClick={this.handleClickButton}
          >
            {label}
          </Button>
        </div>
      </Fragment>
    )
  }

  renderLoading = () => (
    <div className="vtex-minicart__item pa4 flex items-center justify-center relative bg-base">
      <Spinner />
    </div>
  )

  render() {
    const {
      data,
      labelMiniCartEmpty,
      labelButton,
      intl,
      showRemoveButton,
      showDiscount,
      showSku,
      enableQuantitySelector,
      actionOnClick,
      maxQuantity,
      large,
    } = this.props
    const { isUpdating } = this.state

    if (!data || data.loading) {
      return this.renderLoading()
    }

    if (!data.orderForm || !data.orderForm.items.length) {
      const label =
        labelMiniCartEmpty || intl.formatMessage({ id: 'minicart-empty' })
      return this.renderWithoutItems(label)
    }

    const label =
      labelButton || intl.formatMessage({ id: 'finish-shopping-button-label' })
    const labelDiscount = intl.formatMessage({
      id: 'minicart-content-footer-discount',
    })

    return this.renderMiniCartWithItems(
      data.orderForm,
      label,
      labelDiscount,
      showRemoveButton,
      showDiscount,
      showSku,
      enableQuantitySelector,
      maxQuantity,
      actionOnClick,
      isUpdating,
      large
    )
  }
}

export default injectIntl(MiniCartContent)
