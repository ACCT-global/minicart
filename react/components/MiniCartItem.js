import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { FormattedNumber } from 'react-intl'
import { Link } from 'render'

export default class MiniCartItem extends Component {
  static propTypes = {
    imageUrl: PropTypes.string,
    name: PropTypes.string,
    price: PropTypes.number,
    productId: PropTypes.string,
  }

  static contextTypes = {
    culture: PropTypes.object,
  }

  render() {
    const { imageUrl, name, price, productId } = this.props
    return (
      <div className="w-100 pv5 dim">
        <Link
          className="pointer flex flex-row link serious-black f4"
          page={'store/product'}
          params={{ id: productId }}
        >
          <img src={imageUrl} alt={productId} />
          <div>
            <span className="pt6 f4">{name}</span>
            <div className="f5 fw7">
              <FormattedNumber
                value={price}
                style="currency"
                currency={this.context.culture.currency}
                minimumFractionDigits={2}
                maximumFractionDigits={2}
              />
            </div>
          </div>
        </Link>
      </div>
    )
  }
}
