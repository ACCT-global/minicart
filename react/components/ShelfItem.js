import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { FormattedNumber } from 'react-intl'
import { Link } from 'render'

export default class ShelfItem extends Component {
  static propTypes = {
    imageUrl: PropTypes.string,
    name: PropTypes.string,
    price: PropTypes.number,
    productId: PropTypes.string,
  }

  static contextTypes = {
    culture: PropTypes.object,
  }

  trimName = name => {
    return name.length > 2 ? name.substring(0, 18) : name
  }

  render() {
    const { imageUrl, name, price, productId } = this.props
    return (
      <div className="w-100 pv5 dim">
        <Link
          className="pointer flex flex-column link serious-black f4"
          page={'store/product'}
          params={{ id: productId }}
        >
          <img src={imageUrl} alt={name} />
          <span className="pt6 f4">{this.trimName(name)}</span>
          <div className="f2 fw7">
            <FormattedNumber
              value={price}
              style="currency"
              currency={this.context.culture.currency}
              minimumFractionDigits={2}
              maximumFractionDigits={2}
            />
          </div>
        </Link>
      </div>
    )
  }
}
