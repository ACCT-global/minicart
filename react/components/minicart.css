.minicart {
    top: 100%;
    width: 22rem;
}