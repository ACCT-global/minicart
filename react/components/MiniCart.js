import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { graphql, compose } from 'react-apollo'
import Button from '@vtex/styleguide/lib/Button'
import CloseIcon from '@vtex/styleguide/lib/icon/Close'
import orderFormQuery from '../queries/orderFormQuery.gql'
import updateItemsMutation from '../mutations/updateItemsMutation.gql'
import MiniCartItem from './MiniCartItem'

class MiniCart extends Component {
  static propTypes = {
    data: PropTypes.object,
    mutate: PropTypes.func,
  }

  handleRemoveItem = ({ target: { id } }) => {
    const { mutate, data: { orderForm } } = this.props
    const itemPayload = orderForm.items.find(item => item.id === id)
    const index = orderForm.items.indexOf(itemPayload)
    const updatedItem = [itemPayload].map(item => {
      return {
        id: parseInt(item.id),
        index: index,
        quantity: 0,
        seller: 1,
      }
    })

    mutate({
      variables: {
        orderFormId: orderForm.orderFormId,
        items: updatedItem,
      },
      refetchQueries: [{ query: orderFormQuery }],
    })
  }

  getItemId = detailUrl => {
    const regExp = /\/([^)]+)\//
    return regExp.exec(detailUrl)[1]
  }

  render() {
    const { data } = this.props
    if (data.loading) {
      return null
    }
    return (
      <div className="pa3">
        {data.orderForm.items.map(item => (
          <div className="flex flex-row" key={item.id}>
            <MiniCartItem
              imageUrl={item.imageUrl}
              name={item.name}
              price={item.sellingPrice}
              productId={this.getItemId(item.detailUrl)}
            />
            <Button id={item.id} onClick={this.handleRemoveItem}>
              <CloseIcon />
            </Button>
          </div>
        ))}
      </div>
    )
  }
}

export default compose(graphql(orderFormQuery), graphql(updateItemsMutation))(
  MiniCart
)
