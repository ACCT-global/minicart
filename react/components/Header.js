import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { injectIntl, intlShape } from 'react-intl'
import Input from '@vtex/styleguide/lib/Input'
import Button from '@vtex/styleguide/lib/Button'

import MiniCart from './MiniCart'
import CartIcon from '../images/CartIcon'

class Header extends Component {
  constructor(props) {
    super(props)
    this.state = {
      searchValue: '',
      showCart: false,
    }
  }

  static propTypes = {
    name: PropTypes.string,
    intl: intlShape.isRequired,
  }

  translate = id => this.props.intl.formatMessage({ id: `dreamstore.${id}` })

  handleChange = ({ target: { value } }) => {
    this.setState({ searchValue: value })
  }

  handleSearch = () => location.assign(`/${this.state.searchValue}/s`)

  handleCart = () => location.assign('/checkout/#/cart')

  handleCartMouseEnter = () => {
    this.setState({ showCart: true })
  }

  handleMouseLeave = () => {
    this.setState({ showCart: false })
  }

  render() {
    const { account } = global.__RUNTIME__
    const { name } = this.props
    const { searchValue, showCart } = this.state
    return (
      <div className="z-2 flex-ns justify-between items-center w-100 top-0 pa4 pa5-ns bg-white bb bw1 b--serious-black tc tl-ns">
        <a className="link b f3 near-black tc tl-ns" href="/">
          {name || account}
        </a>
        <div className="tr-ns flex items-center">
          <div className="w5-ns mr2">
            <Input
              placeholder={this.translate('search-placeholder')}
              value={searchValue}
              onChange={this.handleChange}
            />
          </div>
          <div className="flex items-center justify-center">
            <Button
              data-test-id="search"
              onClick={this.handleSearch}
              disabled={!searchValue}
            >
              {this.translate('search')}
            </Button>
            <Button
              data-test-id="cart"
              onClick={this.handleCart}
              onMouseEnter={this.handleCartMouseEnter}
            >
              <CartIcon />
            </Button>
            {showCart && (
              <div
                className="absolute z-4 right-0 top-2 w-30 ma6 mt10 bg-white br2 shadow-2"
                onMouseLeave={this.handleMouseLeave}
              >
                <MiniCart />
              </div>
            )}
          </div>
        </div>
      </div>
    )
  }
}

export default injectIntl(Header)
