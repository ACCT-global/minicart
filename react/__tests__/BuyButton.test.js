import React from 'react'
import { shallow } from 'enzyme'
import { mountWithIntl, loadTranslation, setLocale } from 'enzyme-react-intl'
import { BuyButton } from '../components/BuyButton'

describe('BuyButton test', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallow(<BuyButton />)
  })

  it('should render correctly', () => {
    expect(wrapper.getElement()).toMatchSnapshot()
  })

  it('should be a Button element', () => {
    expect(wrapper.is('Button')).toBe(true)
  })

  it('should simulate click', () => {
    loadTranslation('../locales/pt-BR.json')
    setLocale('pt-BR')
    const mutate = jest.fn()
    const onClick = jest.fn()
    const props = { mutate, data: { orderForm: { orderFormId: 1 } }, onClick }
    const wrapper = mountWithIntl(<BuyButton {...props} />)
    const button = wrapper.find('button')
    button.simulate('click')
    expect(mutate.mock.calls.length).toBe(1)
    expect(onClick.mock.calls.length).toBe(1)
  })
})
