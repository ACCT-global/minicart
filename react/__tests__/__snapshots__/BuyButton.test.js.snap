// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BuyButton test should render correctly 1`] = `
<Button
  autoFocus={false}
  block={true}
  disabled={false}
  icon={false}
  large={true}
  onClick={[Function]}
  primary={true}
  secondary={false}
  type="button"
  xLarge={false}
>
  <FormattedMessage
    id="dreamstore.buy"
    values={Object {}}
  />
</Button>
`;
